File: Data Structures & Algorithms/palindromic-substrings/submission-0.js
class Solution {
    /**
     * @param {string} str
     * @return {number}
     */
    countSubstrings(str) {
        /** @type {Map<string, number>} */
        const cache = new Map()
        const build_id = (start, length) => `${start},${length}`
        const is_palindrome = (start, length) => {
            const key = build_id(start, length)
            if (cache.has(key)) return cache.get(key)

            const l = str[start]
            const r = str[start + length]
            const val = length <= 0
                ? true
                : length <= 1
                    ? l === r
                    : l === r && is_palindrome(start + 1, length - 2)
            cache.set(val)
            return val
        }

        let count = 0
        for (let len = 0; len < str.length; len++) {
            for (let start = 0; start + len < str.length; start++) {
                if (is_palindrome(start, len)) count++
            }
        }

        return count
    }
}

File: Data Structures & Algorithms/counting-bits/submission-0.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        def countBit(m: int) -> int:
            count = 0
            while m > 0:
                if m & 1:
                    count += 1
                m >>= 1
            return count

        return [countBit(m) for m in range(n + 1)]
